### PR TITLE
M8: Ledger foundation (schema + domain + repo)

### PR DESCRIPTION
**Pull request by BanerjeeAgniva**
# Pull Request

## Summary

M8 (barcode) **PR1 of 4** — the ledger foundation. Creates the two append-only storage tables that the rest of M8 writes to, plus their JPA entities and repositories. No write path is wired yet (that's PR2–4); this PR only builds the drawer the scan records go in.

**Business anchor:** a ₹1,200 shipment goes missing between the origin-hub dock and the takeoff bag. Six weeks later the station manager must answer *"who physically last touched box `1DD-BLR-260710-000042`, when, and where?"* — unanswerable unless every scan is stored the instant it happens and can never be edited. This PR guarantees the "can never be edited" half.

Design + plan: `docs/M8/M8-BARCODE-DESIGN.md`, `M8-IMPLEMENTATION-PLAN.md`, `M8-DATA-LIFECYCLE.md`.

---

## What Changed

- **`V8_1__create_scan_ledger.sql`** — `scan_ledger`: one immutable row per physical scan. Keyed on `shipment_id` (spine, design D-001); `parcel_id` holds the barcode string (NULL until `LABEL_GENERATED`); `scan_type` is a single column spanning both families (`ScanEventType` ∪ `VanScanType`). Indexed on `(shipment_id, scanned_at)` and `parcel_id`; partial-unique on `client_scan_id` for replay dedup.
- **Append-only enforced at the DB** — a `BEFORE UPDATE OR DELETE` trigger raises, so the audit trail can't be doctored even by raw SQL from another service. Backs up the app-side `updatable=false` mapping.
- **`V8_2__create_parcel_id_counter.sql`** — `parcel_id_counter (hub_iata, day, next_seq)`: per-hub/day sequence for the barcode's `seq6` suffix. Mirrors `orders.shipment_ref_counters`.
- **Entities** — `ScanLedgerEntry` (no setters, all columns `updatable=false`), `ParcelIdCounter` + `ParcelIdCounterId` (`@EmbeddedId`).
- **Repositories** — `ScanLedgerRepository` (the two audit-trail lookups, read+insert only); `ParcelIdCounterRepository` (`insertIfAbsent` + `findByIdWithLock`, same row-lock pattern as the orders counter).
- **`pom.xml`** — added `spring-boot-starter-data-jpa`, `flyway-core`, `postgresql` (runtime), test starter.

---

## Why This Change?

Every other M8 seam depends on this store existing first: M4 state advances off scan events (PR3), routing's `ScanLedgerPort` records van custody here (PR4), and the parcel-ID generator draws from the counter (PR2). Building the schema + immutability guarantee up front means those PRs are thin writers over a proven, append-only foundation rather than each reinventing storage.

---

## Testing
- [x] Tested locally
- [x] Edge cases considered
- [x] No breaking changes introduced

### Test Evidence

`mvn install -pl barcode` (JDK 21) → **BUILD SUCCESS**.

Both migrations were applied inside a **rolled-back transaction** against the shared dev DB (nothing left behind, Flyway untouched — the real checksum-recorded apply happens on next app boot):

```
✅ INSERT works                → rows=1
✅ UPDATE  → ERROR: scan_ledger is append-only: UPDATE is not permitted
✅ DELETE  → ERROR: scan_ledger is append-only: DELETE is not permitted
✅ counter ON CONFLICT DO NOTHING → 2 inserts = 1 row
✅ ROLLBACK — shared DB untouched
```

---

### Areas Affected

- [x] Backend
- [ ] Frontend
- [ ] Routing
- [ ] Infra
- [x] Database
- [ ] NA

### Modules Affected

- [ ] M1 — Auth
- [ ] M2 — Pricing
- [ ] M3 — Grid
- [ ] M4 — Orders
- [ ] M5 — Dispatch
- [ ] M6 — Routing
- [ ] M7 — Hub
- [x] M8 — Barcode
- [ ] M9 — Airline
- [ ] M10 — SLA
- [ ] M11 — Exceptions
- [ ] NA

---

## Risks / Concerns

- **New tables on the shared dev DB.** They apply on the next app boot (Flyway `V8_1`→`V8_2`); both are additive `CREATE TABLE`s in the unused `V8_*` namespace — no collision, no existing-data impact.
- **No runtime writers yet** — tables stay empty until PR2. Intentional; keeps PR1 non-destructive.
- **`parcel_id` NULL on van-scan rows** is expected (routing passes the shipment UUID, not the barcode string, per the documented `v1: parcelId == shipmentId` contract) — the trail is still complete via `shipment_id`.

---

## Checklist
- [x] Code follows project conventions (package layout, no cross-module internal imports)
- [x] Self-review completed
- [x] Append-only audit trail preserved (no mutations to scan/manifest/role-change records)
- [x] No sensitive data or credentials exposed
- [x] Documentation updated if behaviour changed
- [x] Ready for review

### CODE REVIEW

File: barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounter.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * Per-hub, per-day counter backing the seq6 suffix of the parcel barcode (design §3.2).
+ * {@code nextSeq} is the next value to hand out; PR2 increments it under a row lock
+ * (mirrors {@code orders.ShipmentRefCounter}).
+ */
+@Entity
+@Table(name = "parcel_id_counter")
+@Getter
+@Setter
+@NoArgsConstructor
+public class ParcelIdCounter {
```
Comment (BanerjeeAgniva) [inline barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounter.java:21]: What does this do? Give me an example from business perspective. Like what event in real business world will lead to data getting stored here?

Comment (BanerjeeAgniva) [reply barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounter.java:21]: `ParcelIdCounter` is a per-hub, per-day counter, and `nextSeq` is **the next sequence number to stamp onto a parcel's barcode**. A barcode looks like `1DD-DEL-260710-000042` — that trailing `000042` is a value handed out by this counter.

**Real-world event that writes here: a DA prints and sticks a label on a parcel at pickup (`LABEL_GENERATED`).**

Walk it through — first parcel of the day bound for Delhi:
1. DA reaches the customer's door in Bengaluru, taps "Generate Label."
2. The system does `insertIfAbsent('DEL', 2026-07-11)` → creates the row `(DEL, 2026-07-11, next_seq=1)` if it's the first Delhi-bound parcel today.
3. It then reads `next_seq` under a row lock (`findByIdWithLock`), stamps `…-000001` onto the barcode, and bumps `next_seq` to 2.
4. The 2nd Delhi parcel today gets `…-000002`, the 3rd `…-000003`, and so on.

So every label print = one increment of `next_seq` for that (dest-hub, day) pair. The `(hub_iata, day)` key is what guarantees two parcels never collide on the same barcode, and it resets naturally each day (a new date = a fresh row starting at 1).

Note: nothing writes here *yet* — this PR (PR1) only creates the table/entity. The increment logic above lands in **PR2** (parcel-ID generation). Full column-by-column lifecycle is in `docs/M8/M8-DATA-LIFECYCLE.md`.

Comment (BanerjeeAgniva) [reply barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounter.java:21]: why is nextSeq required? Cant we just use an autoincrement implementation on the DB side?

File: barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounterId.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+@Embeddable
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode
+public class ParcelIdCounterId implements Serializable {
+
+    @Column(name = "hub_iata", length = 3, nullable = false)
+    private String hubIata;
+
```
Comment (BanerjeeAgniva) [inline barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounterId.java:23]: What does this do? Give me an example from business perspective. Like what event in real business world will lead to data getting stored here?

Answer in short, try to use / continue with the same example you have given in the above comments

File: barcode/src/main/java/com/oneday/barcode/domain/ScanLedgerEntry.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ * {@code parcelId} barcode string is null until {@code LABEL_GENERATED}. {@code scanType} is a plain
+ * string spanning both families (ScanEventType ∪ VanScanType), so no single Java enum fits it.</p>
+ */
+@Entity
+@Table(name = "scan_ledger")
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class ScanLedgerEntry {
```
Comment (BanerjeeAgniva) [inline barcode/src/main/java/com/oneday/barcode/domain/ScanLedgerEntry.java:31]: What does this do? Give me an example from business perspective. Like what event in real business world will lead to data getting stored here?

Answer in short, try to use / continue with the same example you have given in the above comments

File: barcode/src/main/java/com/oneday/barcode/repository/ParcelIdCounterRepository.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Mirrors {@code orders.ShipmentRefCounterRepository}: {@link #insertIfAbsent} materialises the row
+ * ({@code ON CONFLICT DO NOTHING} for concurrent first-of-day inserts), then {@link #findByIdWithLock}
+ * takes a SELECT FOR UPDATE so the caller reads-and-increments {@code next_seq} atomically in one
+ * round-trip. Caller must be {@code @Transactional} (PR2's parcel-id generator).
+ */
+public interface ParcelIdCounterRepository extends JpaRepository<ParcelIdCounter, ParcelIdCounterId> {
+
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
```
Comment (BanerjeeAgniva) [inline barcode/src/main/java/com/oneday/barcode/repository/ParcelIdCounterRepository.java:24]: What does this do? Give me an example from business perspective. Like what event in real business world will lead to data getting stored here?

Answer in short, try to use / continue with the same example you have given in the above comments

File: barcode/src/main/java/com/oneday/barcode/repository/ScanLedgerRepository.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Deliberately extends the bare {@link Repository} marker, not {@code JpaRepository}: the ledger is
+ * append-only (V8_1 trigger backstops it), so no {@code delete*}/bulk-update method is exposed at all —
+ * only insert ({@link #save}) and the "who touched this box, when, where" trail lookups. The DB
+ * trigger is the last line of defence; this interface is the first (mutation isn't even reachable).
+ */
+public interface ScanLedgerRepository extends Repository<ScanLedgerEntry, UUID> {
```
Comment (BanerjeeAgniva) [inline barcode/src/main/java/com/oneday/barcode/repository/ScanLedgerRepository.java:15]: What does this do? Give me an example from business perspective. Like what event in real business world will lead to data getting stored here?

Answer in short, try to use / continue with the same example you have given in the above comments

File: barcode/src/main/resources/db/migration/V8_1__create_scan_ledger.sql [hunk 6]
```diff
@@ -0,0 +1,40 @@
+-- M8 §4.1 — the append-only scan ledger: one immutable row per physical scan of a parcel, forever.
+-- The single source of truth for "who touched this box, when, where" (design D-001: keyed on
+-- shipment_id — the same UUID routing/hub already pass; the human barcode string in parcel_id is
+-- layered on from LABEL_GENERATED onward). Two write paths feed it (design §4.3): REST lifecycle
+-- scans (M8 PR3) and the in-process van custody port (M8 PR4). Never updated, never deleted.
+CREATE TABLE scan_ledger (
```
Comment (BanerjeeAgniva) [inline barcode/src/main/resources/db/migration/V8_1__create_scan_ledger.sql:6]: What does this do? Give me an example from business perspective. Like what event in real business world will lead to data getting stored here?

Answer in short, try to use / continue with the same example you have given in the above comments

File: barcode/src/main/resources/db/migration/V8_2__create_parcel_id_counter.sql [hunk 7]
```diff
@@ -0,0 +1,10 @@
+-- M8 §3.2 — per-hub, per-day sequence backing the seq6 suffix of the parcel barcode
+-- (1DD-{destHubIATA}-{yyMMdd}-{seq6}). Same row-lock pattern as orders.shipment_ref_counters:
+-- INSERT ... ON CONFLICT DO NOTHING to materialise the row, then SELECT FOR UPDATE to increment
+-- (M8 PR2 owns the generation logic; this PR only provides the table).
+CREATE TABLE parcel_id_counter (
```
Comment (BanerjeeAgniva) [inline barcode/src/main/resources/db/migration/V8_2__create_parcel_id_counter.sql:5]: What does this do? Give me an example from business perspective. Like what event in real business world will lead to data getting stored here?

Answer in short, try to use / continue with the same example you have given in the above comments

File: barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounter.java (hunk 1 above)
Comment (BanerjeeAgniva) [reply barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounter.java:21]: Good question. A DB autoincrement (SERIAL / IDENTITY / a Postgres sequence) gives **one global, ever-growing** number. Our barcode needs the opposite: `1DD-DEL-260711-000042`, where `000042` means *the 42nd Delhi-bound parcel **today***, scoped to `(hub, day)` and reset every midnight.

A global sequence can't do that:
- it never resets per day (you'd see `…-000042` on day 1 and `…-018342` on day 90);
- it isn't per-hub (BLR and DEL couldn't both sit at `000001` on the same day);
- it leaks total platform volume in every printed barcode.

You *could* `CREATE SEQUENCE` dynamically per hub-per-day, but that's unbounded sequence sprawl (a new sequence object × 5 hubs × 365 days). A single keyed counter row `(hub_iata, day, next_seq)` is the clean way — and it mirrors the already-proven `orders.shipment_ref_counters`. That's why `next_seq` is app-managed under a row lock rather than a DB autoincrement.

File: barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounterId.java (hunk 2 above)
Comment (BanerjeeAgniva) [reply barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounterId.java:23]: Short: it's the **key** of the counter — `(hub_iata='DEL', day=2026-07-11)`. Continuing the example: when Priya's Delhi-bound parcel gets its label today, this key points at "the Delhi counter for today." It's what makes the running number **per-hub, per-day** — DEL and BLR each keep their own count, and each resets at midnight (a new date = a fresh key starting at 1).

File: barcode/src/main/java/com/oneday/barcode/domain/ScanLedgerEntry.java (hunk 3 above)
Comment (BanerjeeAgniva) [reply barcode/src/main/java/com/oneday/barcode/domain/ScanLedgerEntry.java:31]: Short: one **immutable row = one physical scan** of a parcel. Continuing the example: when the Bengaluru hub gun scans Priya's box at the dock (`HUB_ORIGIN_IN`), one row is written here — `shipment_id=S`, `scan_type=HUB_ORIGIN_IN`, `location_type=HUB`, `scanned_at=now`. Every scan on the box's journey (label → van → hub → airport → …) appends one such row; none are ever edited.

File: barcode/src/main/java/com/oneday/barcode/repository/ParcelIdCounterRepository.java (hunk 4 above)
Comment (BanerjeeAgniva) [reply barcode/src/main/java/com/oneday/barcode/repository/ParcelIdCounterRepository.java:24]: Short: it's **how PR2 talks to the counter**. Continuing the example, at Priya's label print: `insertIfAbsent('DEL', 2026-07-11)` creates the Delhi-today row if it's the first such parcel, then `findByIdWithLock` locks + reads it to grab `000042` and bump it to `000043` — atomically, so two DAs printing at the same instant can't be handed the same number.

File: barcode/src/main/java/com/oneday/barcode/repository/ScanLedgerRepository.java (hunk 5 above)
Comment (BanerjeeAgniva) [reply barcode/src/main/java/com/oneday/barcode/repository/ScanLedgerRepository.java:15]: Short: the **read side** of the ledger (writes happen via `save` at each scan). Continuing the example: six weeks later Priya's parcel is reported lost — the station manager calls `findByParcelIdOrderByScannedAtAsc('1DD-DEL-260711-000042')` and gets the full scan trail in order (door → pickup van → BLR hub → airport → …), answering "who last touched this box, when, where."

File: barcode/src/main/resources/db/migration/V8_1__create_scan_ledger.sql (hunk 6 above)
Comment (BanerjeeAgniva) [reply barcode/src/main/resources/db/migration/V8_1__create_scan_ledger.sql:6]: Short: this creates the **ledger table itself** — the append-only book of every scan. Continuing the example: the first row for Priya's parcel lands when the DA prints its label at her door (`LABEL_GENERATED`); every later scan (van, hub, airport…) appends another row. The trigger below makes rows un-editable, so the audit trail can't be doctored.

File: barcode/src/main/resources/db/migration/V8_2__create_parcel_id_counter.sql (hunk 7 above)
Comment (BanerjeeAgniva) [reply barcode/src/main/resources/db/migration/V8_2__create_parcel_id_counter.sql:5]: Short: this creates the **counter table** that hands out the `000042` in the barcode. Continuing the example: when Delhi-bound parcels get labels today, the row `(DEL, 2026-07-11, next_seq)` is created and its `next_seq` ticks up per label — 1, 2, 3… — resetting tomorrow with a fresh `(DEL, 2026-07-12)` row.

File: barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounterId.java (hunk 2 above)
Comment (BanerjeeAgniva) [reply barcode/src/main/java/com/oneday/barcode/domain/ParcelIdCounterId.java:23]: hubIata = the 3-letter IATA code of a hub's city — the same standardized airport codes the airline industry uses (IATA = International Air Transport Association).
